sql/core/src/test/scala/org/apache/spark/sql/JoinSuite.scala

```diff
@@ -1188,4 +1188,42 @@ class JoinSuite extends QueryTest with SharedSparkSession with AdaptiveSparkPlan
         classOf[BroadcastNestedLoopJoinExec]))
     }
   }
+
+  test("SPARK-32399: Full outer shuffled hash join") {
```

Comment:
I am curious if it might be possible to unconditionally enable this optimization for _all_ full outer joins to make sure that this optimization unconditionally works ? A similar approach was taken in PR #29104, if you search for `CONFIG_DIM1 spark.sql.optimizeNullAwareAntiJoin=true` in all `.sql` files like `group-by-filter.sql` for example.

Reply:
@agrawaldevesh - I actually found [there are existing tests for comparing BHJ/SHJ/SMJ already in that test suite](https://github.com/apache/spark/blob/master/sql/core/src/test/resources/sql-tests/inputs/subquery/in-subquery/in-joins.sql#L8-L10). But the config value for SHJ is wrong which should not be spark.sql.autoBroadcastJoinThreshold=-1. Created a followup JIRA (https://issues.apache.org/jira/browse/SPARK-32577) to fix it.

Reply:
I didn't fully follow: Do you mean that until https://issues.apache.org/jira/browse/SPARK-32577 is fixed we don't have very high confidence in this optimization truly producing the same results as without it ?

All I am trying to ascertain is whether this optimization is safe in all cases: Would the results produced for full outer join be identical both with and without this optimization ? My understanding is that currently, this is only validated by the above scala unit tests that have been newly added, but it hasn't been fully validated for all full-outer-join scenarios due to https://issues.apache.org/jira/browse/SPARK-32577. Is that an accurate understanding ?

Reply:
> Do you mean that until https://issues.apache.org/jira/browse/SPARK-32577 is fixed we don't have very high confidence in this optimization truly producing the same results as without it ?

No. I am saying that there are existing tests for comparing BHJ/SHJ/SMJ per my last comment link, but it does not trigger SHJ as expected due to wrong config. But that can be fixed separately as that is to fix a regression in unit test.

To enable/disable SHJ for bulk of test queries is not very easy, as SHJ depending on [carefully tuning of threshold config](https://github.com/apache/spark/pull/29342#discussion_r467167531), there's no way to disable/enable SHJ for a query with true/false config. So to validate SHJ for all test queries under `sql-tests`, I need to tune the config for every join queries, which I think it would involve too much work. In addition, I noticed for newly added features, we rarely use `sql-tests` for validation by enabling/disabling feature, but mostly adding unit test. Wondering how do you think the added unit test in `JoinSuite.scala`?

Reply:
Thanks for noticing the lack in the test coverage and explaining this to me ! I understand this better now.

But you see where I am coming from :-) I am simply concerned about data corruption (more precisely wrong result computation). 

Ideally, we should first fix the test, land that fix and then retest this PR with the tests fixed. It seems like it shouldn't be too hard, nor cause any merge conflicts, but it would delay landing this PR.

As a compromise, may I please suggest that:
- Make that test change as mentioned in SPARK-32577 locally and run the sql tests locally to verify that they work.
- Run JoinSuite with code coverage enabled and ensure that all of the changed code paths in this PR are covered. 

I recently worked on a join related feature and I was humbled by how many bugs I was able to make around nulls, strings and its relationship with downstream union queries :-D

I think the tests in JoinSuite are good, but I am more concerned about the unknown unknowns and any regressions. And interactions of join with other operators like aggregations and unions.

This is not a blocker at all !. Just due diligence and paranoia.